nexus/src/app/sagas/instance_update/destroyed.rs

```diff
@@ -0,0 +1,256 @@
+    let osagactx = sagactx.user_data();
+    let RealParams { ref serialized_authn, ref authz_instance, .. } =
+        sagactx.saga_params::<RealParams>()?;
+    let vmm_id = sagactx.lookup::<PropolisUuid>(DESTROYED_VMM_ID)?;
+
+    let opctx =
+        crate::context::op_context_for_saga_action(&sagactx, serialized_authn);
+
+    info!(
+        osagactx.log(),
+        "instance update (active VMM destroyed): deallocating sled resource reservation";
+        "instance_id" => %authz_instance.id(),
+        "propolis_id" => %vmm_id,
+        "instance_update" => %"VMM destroyed",
+    );
+
+    osagactx
+        .datastore()
+        .sled_reservation_delete(&opctx, vmm_id.into_untyped_uuid())
+        .await
+        .or_else(|err| {
+            // Necessary for idempotency
+            match err {
+                Error::ObjectNotFound { .. } => Ok(()),
+                _ => Err(err),
+            }
+        })
+        .map_err(ActionError::action_failed)
+}
+
+pub(super) async fn siu_destroyed_release_virtual_provisioning(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let osagactx = sagactx.user_data();
+    let RealParams { ref serialized_authn, ref authz_instance, state, .. } =
+        sagactx.saga_params::<RealParams>()?;
+
+    let vmm_id = sagactx.lookup::<PropolisUuid>(DESTROYED_VMM_ID)?;
+    let instance = state.instance;
+    let instance_id = InstanceUuid::from_untyped_uuid(authz_instance.id());
+
+    let opctx =
+        crate::context::op_context_for_saga_action(&sagactx, serialized_authn);
+
+    // `virtual_provisioning_collection_delete_instace` will only delete virtual
+    // provisioning records that are *less than* the max generation parameter,
+    // not less than or equal to it --- the idea is that the generation number
+    // has already been advanced when we are deallocating the virtual
+    // provisioning records. This is kind of an artifact of sled-agent
+    // previously owning instance runtime state generations, since the
+    // sled-agent would have already advanced the instance's generation.
+    //
+    // However, now that the instance record is owned by Nexus, and we are
+    // updating the instance in response to a VMM state update from sled-agent,
+    // the instance record snapshot we are holding has not yet had its
+    // generation advanced, so we want to allow deleting virtual provisioning
+    // records that were created with the instance's current generation. The
+    // generation will be advanced at the end of this saga, once we have updated
+    // the actual instance record.
+    let max_gen = instance.runtime_state.gen.next();
+    let result = osagactx
+        .datastore()
+        .virtual_provisioning_collection_delete_instance(
+            &opctx,
+            instance_id,
+            instance.project_id,
+            i64::from(instance.ncpus.0 .0),
+            instance.memory,
+            i64::try_from(&max_gen).unwrap(),
```

Comment:
I think the generation number check here should just go away entirely. We've discussed its purpose before, but as a reminder to future generations who might look at this PR, it was meant to prevent the following race:

- sled agent sends an "instance stopped now" instance/VMM update to Nexus
- Nexus doesn't respond for a while
- sled agent's client times out and reissues the update to a different Nexus
- second Nexus processes the request, removes provisioning counters, and marks the instance as stopped
- instance starts again
- previous Nexus request (the one that sled agent abandoned) wakes up and removes provisioning counters from a running instance. Oops.

Checking the instance generation number (which was bumped when the instance stop was recorded) prevents this by allowing the resuscitated Nexus call to see that another Nexus already took the action that it was going to take.

With the update saga, though, we don't have concurrent actors trying to remove counters on a restarted instance--by definition there's only one updater at once, and it has to remove counters and write back an updated instance record before an instance can start again. So I bet we can just remove the generation number check entirely.

(The other user of this function is the instance start unwind path, which has a similar property: another attempt to start the instance can't overtake the saga's attempt to release counters, because an instance can have only one start saga at once, and it has to release counters before another start saga can proceed.)

Reply:
Yeah, thanks for the reminder --- I had planned to go back and remove this but was going to wait until I had gotten the complete saga sketched out. Thanks for the reminder, I had kinda forgotten.

Reply:
Got rid of it in 877074b8ca19d9ddcb5a04f1ffb819c4d635871d.